File: app/javascript/channels/room_channel.js
// このファイルは「WebSocket処理のクライアントサイド側」を受け持っています

import consumer from "./consumer"
document.addEventListener('turbolinks:load', () => {
  const channel = "RoomChannel";
  const data = document.getElementById('data').dataset;
  const chatroomId = data.chatroomId;
  const userId = data.userId;

  consumer.subscriptions.create(
    {
      channel: channel, chatroom_id: chatroomId, user_id: userId
    },
    {
      // 接続時のコールバックメソッド
      connected() {


        // 投稿フォームのメッセージ内容と送信ボタンを取得
        const chatFormText = document.querySelector('textarea[data-behavior="chatFormText"]');
        const chatFormRoomId = document.querySelector('input[data-behavior="chatFormRoomId"]');
        const chatFormSubmit = document.querySelector('input[data-behavior="chatFormSubmit"]');
        // セレクターにidを使わない理由はchatGPTによると・・・
        //   idを使って要素を特定する場合は、同じidを複数の要素に使うことができないため、同じ機能を持った要素を追加する場合に不便です。
        //   一方、data属性を使ったセレクターを使うことで、同じ機能を持った要素を区別することができます。
        //   また、idを使ったセレクターは他のJavaScriptライブラリとの競合を引き起こす可能性があるため、避けるべきです。
        //   したがって、data属性を使ったセレクターを使うことが推奨されます。


        // 送信ボタンがクリックされた時
        chatFormSubmit.addEventListener('click', (event) => {
          // speakメソッドにフォームの内容とroom_idを渡す（サーバーサイドへ送信）
          this.speak(chatFormText.value, chatFormRoomId.value);
          // フォームの文字を消去
          chatFormText.value = '';
          // 通常のフォーム送信を防止
          return event.preventDefault();
        });
      },

      // 切断時のコールバックメソッド
      disconnected() {
        // Called when the subscription has been terminated by the server
      },

      // サーバーサイドからのデータを受信するメソッド
      received(data) {
        const element = document.querySelector('#messages')
        if (userId == data['user_id']) {
          element.insertAdjacentHTML('beforeend',
          `<div class="row">
          <div class="col"></div>
          <div class="col text-right">
          <small>
          ${data['user_name']}
          ${data['created_at']}
          </small>
          <div class="bg-info rounded m-1 p-1">
          <p>${data['message']}</p>
          </div>
          </div>
          </div>`
          )
        } else {
          element.insertAdjacentHTML('beforeend',
          `<div class="row">
          <div class="col">
          <small>
          ${data['user_name']}
          ${data['created_at']}
          </small>
          <div class="bg-secondary rounded m-1 p-1">
          <p>${data['message']}</p>
          </div>
          </div>
          <div class="col"></div>
          </div>`
          )
        }
      },

      // サーバーサイドへメッセージを送信するメソッド
      speak: function(message, room_id) {
        // this.perform('speak') はサーバーサイドのspeakメソッドをWebSocket通信経由で呼び出します
        // performメソッドを使ってメッセージを送信します
        return this.perform('speak', {message: message, room_id: room_id});
      }
    }
  );
});

File: app/channels/room_channel.rb
# このファイルを「チャネル」と呼び、「WebSocket処理のサーバーサイド」を受け持っています
# （クライアント側は app/javascript/channels/room_channel.js）
# チャネルはWebsocket通信におけるコントローラのような役割を果たします
# クライアントはWebsocket通信を通じてチャネルと関連付けされる＝「購読（subscription）」
# 購読は一度に複数のチャネルに対して作ることができ、たとえばチャットルーム用のチャネルと、通知欄用のチャネルを同時に購読するようなことができます

class RoomChannel < ApplicationCable::Channel

  # 購読後に呼び出されるメソッド
  def subscribed
    # 「room_channel」というストリーム名を設定しています
    # 接続しているクライアントへメッセージを送る（ブロードキャスト）時にこのストリーム名を使用します
    # ここではspeakメソッドでブロードキャストする時に使っています
    @user = User.find(params[:user_id])
    reject if @user.nil?
    @chatroom = Room.find(params[:chatroom_id])
    reject if @chatroom.nil?
    stream_from "room_channel_#{params[:chatroom_id]}"
  end

  # 購読解除後に呼び出されるメソッド
  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  # 購読しているクライアントから呼び出された時に実行されるメソッド
  def speak(data)
    # クライアント側から送られてきたデータ(data)でメッセージのレコードを作ります
    message = Message.create!(message: data["message"], user_id: params[:user_id], room_id: data["room_id"])
    # チャネルを購読している人へブロードキャスト！
    ActionCable.server.broadcast(
      # "room_channel_#{data["room_id"]}", { message: data["message"] }
      "room_channel_#{message.room_id}", { message: render_message(message) }
    )
  end

  private

  def render_message(message)
    # ApplicationController.renderを使うと、コントローラ外からテンプレートのレンダリングを行うことができる
    ApplicationController.render_with_signed_in_user(
      message.user,
      partial: "public/messages/message",
      locals: { message: message }
    )
  end
end
